<h1 style="text-align:center; font-size:42px; margin-top:40px;">
Exploration Initiale du Stack Overflow Developer Survey 2024
</h1>



## 1. Contexte et objectif

Ce module constitue la première étape de l’exploration du dataset Stack Overflow Developer Survey 2024.  
L’objectif est d’obtenir une vue globale de la structure des données, d’identifier les colonnes clés, de mesurer la qualité du dataset et de préparer la phase de nettoyage.

Cette exploration vise à :
- analyser la structure générale du fichier (dimensions, types, granularité) ;
- détecter les premières anomalies (valeurs manquantes, colonnes inutiles, codifications incohérentes) ;
- évaluer la pertinence analytique des variables ;
- dégager des premiers insights descriptifs sur les répondants.

Cette étape pose la base du pipeline de préparation et conditionne les transformations à venir.


## 2. Import des librairies + chargement du dataset

In [ ]:
# Téléchargement du DataSet en local
#!curl -o ../Data/Survey/raw/survey-data_V0.csv https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/VYPrOu0Vs3I0hKLLjiPGrA/survey-data-with-duplicate.csv

# ou bein decompresser survey-data_V0.7z existant dans ""../Data/Survey/raw"


In [ ]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 60)

# Charger dataset brut
df = pd.read_csv("../Data/Survey/raw/survey-data_V0.csv")

# Afficher les 5 premieres lignes
df.head()


,ResponseId,MainBranch,Age,Employment,RemoteWork,Check,CodingActivities,EdLevel,LearnCode,LearnCodeOnline,TechDoc,YearsCode,YearsCodePro,DevType,OrgSize,PurchaseInfluence,BuyNewTool,BuildvsBuy,TechEndorse,Country,Currency,CompTotal,LanguageHaveWorkedWith,LanguageWantToWorkWith,LanguageAdmired,DatabaseHaveWorkedWith,DatabaseWantToWorkWith,DatabaseAdmired,PlatformHaveWorkedWith,PlatformWantToWorkWith,PlatformAdmired,WebframeHaveWorkedWith,WebframeWantToWorkWith,WebframeAdmired,EmbeddedHaveWorkedWith,EmbeddedWantToWorkWith,EmbeddedAdmired,MiscTechHaveWorkedWith,MiscTechWantToWorkWith,MiscTechAdmired,ToolsTechHaveWorkedWith,ToolsTechWantToWorkWith,ToolsTechAdmired,NEWCollabToolsHaveWorkedWith,NEWCollabToolsWantToWorkWith,NEWCollabToolsAdmired,OpSysPersonal use,OpSysProfessional use,OfficeStackAsyncHaveWorkedWith,OfficeStackAsyncWantToWorkWith,OfficeStackAsyncAdmired,OfficeStackSyncHaveWorkedWith,OfficeStackSyncWantToWorkWith,OfficeStackSyncAdmired,AISearchDevHaveWorkedWith,AISearchDevWantToWorkWith,AISearchDevAdmired,NEWSOSites,SOVisitFreq,SOAccount,SOPartFreq,SOHow,SOComm,AISelect,AISent,AIBen,AIAcc,AIComplex,AIToolCurrently Using,AIToolInterested in Using,AIToolNot interested in Using,AINextMuch more integrated,AINextNo change,AINextMore integrated,AINextLess integrated,AINextMuch less integrated,AIThreat,AIEthics,AIChallenges,TBranch,ICorPM,WorkExp,Knowledge_1,Knowledge_2,Knowledge_3,Knowledge_4,Knowledge_5,Knowledge_6,Knowledge_7,Knowledge_8,Knowledge_9,Frequency_1,Frequency_2,Frequency_3,TimeSearching,TimeAnswering,Frustration,ProfessionalTech,ProfessionalCloud,ProfessionalQuestion,Industry,JobSatPoints_1,JobSatPoints_4,JobSatPoints_5,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,SurveyLength,SurveyEase,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,Under 18 years old,"Employed, full-time",Remote,Apples,Hobby,Primary/elementary school,Books / Physical media,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,United States of America,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,I have never visited Stack Overflow or the Sta...,NaN,NaN,NaN,NaN,NaN,Yes,Very favorable,Increase productivity,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,No,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,I am a developer by profession,35-44 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,API document(s) and/or SDK document(s);User gu...,20,17,"Developer, full-stack",NaN,NaN,NaN,NaN,NaN,United Kingdom of Great Britain and Northern I...,NaN,NaN,Bash/Shell (all shells);Go;HTML/CSS;Java;JavaS...,Bash/Shell (all shells);Go;HTML/CSS;Java;JavaS...,Bash/Shell (all shells);Go;HTML/CSS;Java;JavaS...,Dynamodb;MongoDB;PostgreSQL,PostgreSQL,PostgreSQL,Amazon Web Services (AWS);Heroku;Netlify,Amazon Web Services (AWS);Heroku;Netlify,Amazon Web Services (AWS);Heroku;Netlify,Express;Next.js;Node.js;React,Express;Htmx;Node.js;React;Remix,Express;Node.js;React,NaN,NaN,NaN,NaN,NaN,NaN,Docker;Homebrew;Kubernetes;npm;Vite;Webpack,Docker;Homebrew;Kubernetes;npm;Vite;Webpack,Docker;Homebrew;Kubernetes;npm;Vite;Webpack,PyCharm;Visual Studio Code;WebStorm,PyCharm;Visual Studio Code;WebStorm,PyCharm;Visual Studio Code;WebStorm,MacOS;Windows,MacOS,NaN,NaN,NaN,Microsoft Teams;Slack,Slack,Slack,NaN,NaN,NaN,Stack Overflow for Teams (private knowledge sh...,Multiple times per day,Yes,Multiple times per day,Quickly finding code solutions;Finding reliabl...,"Yes, definitely","No, and I don't plan to",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Yes,Individual contributor,17.0,Agree,Disagree,Agree,Agree,Agree,Neither agree nor disa

## 3. Dimensions et structure du dataset


In [21]:
#(nombre de lignes, nombre colonnes)
df.shape

(65457, 114)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65457 entries, 0 to 65456
Columns: 114 entries, ResponseId to JobSat
dtypes: float64(13), int64(1), object(100)
memory usage: 56.9+ MB


In [23]:
# Liste des colonnes
df.columns.to_list()

['ResponseId',
 'MainBranch',
 'Age',
 'Employment',
 'RemoteWork',
 'Check',
 'CodingActivities',
 'EdLevel',
 'LearnCode',
 'LearnCodeOnline',
 'TechDoc',
 'YearsCode',
 'YearsCodePro',
 'DevType',
 'OrgSize',
 'PurchaseInfluence',
 'BuyNewTool',
 'BuildvsBuy',
 'TechEndorse',
 'Country',
 'Currency',
 'CompTotal',
 'LanguageHaveWorkedWith',
 'LanguageWantToWorkWith',
 'LanguageAdmired',
 'DatabaseHaveWorkedWith',
 'DatabaseWantToWorkWith',
 'DatabaseAdmired',
 'PlatformHaveWorkedWith',
 'PlatformWantToWorkWith',
 'PlatformAdmired',
 'WebframeHaveWorkedWith',
 'WebframeWantToWorkWith',
 'WebframeAdmired',
 'EmbeddedHaveWorkedWith',
 'EmbeddedWantToWorkWith',
 'EmbeddedAdmired',
 'MiscTechHaveWorkedWith',
 'MiscTechWantToWorkWith',
 'MiscTechAdmired',
 'ToolsTechHaveWorkedWith',
 'ToolsTechWantToWorkWith',
 'ToolsTechAdmired',
 'NEWCollabToolsHaveWorkedWith',
 'NEWCollabToolsWantToWorkWith',
 'NEWCollabToolsAdmired',
 'OpSysPersonal use',
 'OpSysProfessional use',
 'OfficeStackAsyncHa

## 4. Observations initiales (résumé)

- 65 457 lignes et 114 colonnes → dataset large et structuré en multiples thématiques.
- Majorité de colonnes textuelles (`object`), typique d’un survey à choix multiples.
- Colonnes multivaluées nombreuses (langages, outils, bases de données, plateformes), nécessitant un traitement `split` puis `explode`.
- Variables thématiques distinctes : profil, stack technologique, IA, comportement, satisfaction, compensation.
- Présence de colonnes numériques clés (`ConvertedCompYearly`, `YearsCodePro`) à vérifier pour outliers et formats.
- Le dataset exige une phase de préparation structurée : nettoyage, normalisation, reformatage des blocs multivalués.


## 5. Analyse des valeurs manquantes

In [ ]:
# % des aleurs manquantes
missing = df.isna().mean().round(2).sort_values(ascending=False)
missing


AINextMuch less integrated    0.98
AINextLess integrated         0.96
AINextNo change               0.81
AINextMuch more integrated    0.79
EmbeddedAdmired               0.74
                              ... 
MainBranch                    0.00
Check                         0.00
Employment                    0.00
Age                           0.00
ResponseId                    0.00
Length: 114, dtype: float64

### Observation

- Certaines colonnes présentent un niveau élevé de valeurs manquantes (>40%), typique dans les questionnaires facultatifs.
- Ces colonnes seront soit :
  - retirées (si non pertinentes pour le projet),
  - imputées avec une catégorie “Unknown”,
  - ou traitées comme variables binaires (présence / absence d'information).

Cette étape guide la stratégie de nettoyage dans le notebook suivant.


## 6. Détection des doublons

In [25]:
df.duplicated().sum()

np.int64(20)

### Observation
- Les doublons éventuels seront supprimés dans le module de cleaning.  

## 7. Exploration des colonnes clés (colonnes technologiques)

In [26]:
df[['LanguageHaveWorkedWith', 'DatabaseHaveWorkedWith','PlatformHaveWorkedWith','WebframeHaveWorkedWith','EmbeddedHaveWorkedWith','MiscTechHaveWorkedWith']].head(3)

,LanguageHaveWorkedWith,DatabaseHaveWorkedWith,PlatformHaveWorkedWith,WebframeHaveWorkedWith,EmbeddedHaveWorkedWith,MiscTechHaveWorkedWith
0,NaN,NaN,NaN,NaN,NaN,NaN
1,Bash/Shell (all shells);Go;HTML/CSS;Java;JavaS...,Dynamodb;MongoDB;PostgreSQL,Amazon Web Services (AWS);Heroku;Netlify,Express;Next.js;Node.js;React,NaN,NaN
2,C#,Firebase Realtime Database,Google Cloud,ASP.NET CORE,Rasberry Pi,.NET (5+) ;.NET Framework (1.0 - 4.8);.NET MAUI


### Observation

- Les colonnes technologiques sont multivaluées, séparées par des `;`.  
- Une transformation en structure longue (explosion) sera nécessaire dans l’EDA.


## 8. Colonnes inutiles

Seules les colonnes utiles à l’analyse seront conservées.  
Les colonnes administratives, trop incomplètes ou sans valeur analytique directe seront retirées. Cela inclut : les variables avec >90 % de valeurs manquantes, ainsi que les séries très spécifiques (`Knowledge_*`, `Frequency_*`, `JobSatPoints_*`).  
L’objectif est de travailler uniquement sur les variables pertinentes pour répondre aux questions analytiques du projet.


## 9.  Résumé de l'exploration initiale

Cette première exploration permet d’établir une vision claire du dataset et de définir les priorités pour la phase de nettoyage :

- structure volumineuse, majoritairement textuelle et multivaluée ;
- présence de valeurs manquantes importantes sur certains champs ;
- variables clés identifiées pour l'analyse : langages, bases de données, expérience, statut professionnel, localisation, revenus ;
- besoin d’un traitement spécifique pour les champs multivalués (explosion, normalisation) ;
- présence de doublons ;

Cette étape fournit la base nécessaire pour les notebooks suivants consacrés au **cleaning structuré et à la préparation analytique du dataset**.


<!--
## Change Log
|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2024-10-29|3.0|Madhusudhan|Updated lab|
|2024-09-23|2.0|Madhusudhan|Updated lab|
|2024-09-22|1.0|Raghul Ramesh|Created lab|
--!>
